mathics/core/atoms.py

```diff
@@ -1062,6 +1067,114 @@ def is_zero(self) -> bool:
+        "Integer8": numpy.dtype("int8"),
+        "Integer16": numpy.dtype("int16"),
+        "Integer32": numpy.dtype("int32"),
+        "Integer64": numpy.dtype("int64"),
+        "Real32": numpy.dtype("float32"),
+        "Real64": numpy.dtype("float64"),
+        "ComplexReal32": numpy.dtype("complex64"),
+        "ComplexReal64": numpy.dtype("complex128"),
+    }
+    NUMERIC_ARRAY_DTYPE_TO_NAME = {
+        dtype: name for name, dtype in NUMERIC_ARRAY_TYPE_MAP.items()
+    }
+else:  # pragma: no cover - executed only when numpy is absent
+    NUMERIC_ARRAY_TYPE_MAP = {}
+    NUMERIC_ARRAY_DTYPE_TO_NAME = {}
+
+
+# TODO: would it be useful to follow the example of Complex and parameterize by type?
+# would that be array.dtype or the MMA type from the map above?
+class NumericArray(Atom, ImmutableValueMixin):
+    """
+    NumericArray provides compact storage and efficient access for machine-precision numeric arrays,
+    backed by NumPy arrays.
+    """
+
+    class_head_name = "NumericArray"
+
+    def __init__(self, value, dtype=None):
+        if numpy is None:
+            raise ImportError("numpy is required for NumericArray")
+
+        # compute value
+        if not isinstance(value, numpy.ndarray):
+            value = numpy.asarray(value, dtype=dtype)
+        elif dtype is not None:
+            value = value.astype(dtype)
+        self.value = value
+
+        # check type
+        self._type_name = NUMERIC_ARRAY_DTYPE_TO_NAME.get(self.value.dtype, None)
+        if not self._type_name:
+            allowed = ", ".join(str(dtype) for dtype in NUMERIC_ARRAY_TYPE_MAP.values())
+            message = f"Argument 'value' must be one of {allowed}; is {str(self.value.dtype)}."
+            raise ValueError(message)
+
+        # summary and hash
+        self._summary = (self._type_name, self.value.shape, self.value.tobytes())
+        shape_string = "×".join(str(dim) for dim in self.value.shape) or "0"
+        self._summary_string = f"{self._type_name}, {shape_string}"
+        self._hash = None
+
+    # TODO: this is potentially expensive - what if we left it unimplemented? is hashing a numpy array reasonable?
+    # TODO: or maybe make self._summary included only some of the bytes, since it's just a hash?
+    def __hash__(self):
+        if not self._hash:
+            print("HASHING NUMERICARRAY")
+            self._hash = hash(("NumericArray", self._summary))
+        return self._hash
+
+    def __str__(self) -> str:
+        return f"NumericArray[{self._summary_string}]"
+
+    def atom_to_boxes(self, f, evaluation):
+        return String(f"<{self._summary_string}>")
+
+    def do_copy(self) -> "NumericArray":
+        return NumericArray(self.value.copy())
+
+    def default_format(self, evaluation, form) -> str:
+        return f"NumericArray[<{self._summary_string}>]"
+
+    @property
+    def element_order(self) -> tuple:
+        return (BASIC_ATOM_STRING_OR_BYTEARRAY_SORT_KEY, *self._summary)
```

Comment:
The name `BASIC_ATOM_SRING_OR_BYTEARRAY_SORT_KEY` is changed in https://github.com/Mathics3/mathics-core/pull/1505 so this will need to be adjust.

But the _reason_ this is not correct is that, in Python, you can't compare a NumericArray with a String, even though they have similar structure (have elements). 

Exprementation using the WMA function `Order[]` can be used to find out which of these comes (somewhat arbitrarily) first.  And then how that compares with ByteArray, and this then gets set in `mathics.core.keycomparable`.